model/map/src/main/java/org/keycloak/models/map/storage/criteria/ModelCriteriaNode.java

```diff
@@ -35,45 +36,45 @@
 
     public static enum ExtOperator {
         AND {
-            @Override public <M, C extends ModelCriteriaBuilder<M>> C apply(C mcb, ModelCriteriaNode<M> node) {
+            @Override public <M, C extends ModelCriteriaBuilder<M, C>> C apply(C mcb, ModelCriteriaNode<M> node) {
                 if (node.getChildren().isEmpty()) {
                     return null;
                 }
-                final ModelCriteriaBuilder[] operands = node.getChildren().stream()
+                final C[] operands = node.getChildren().stream()
                   .map(n -> n.flashToModelCriteriaBuilder(mcb))
                   .filter(Objects::nonNull)
-                  .toArray(ModelCriteriaBuilder[]::new);
-                return operands.length == 0 ? null : (C) mcb.and(operands);
+                  .toArray(n -> (C[]) Array.newInstance(mcb.getClass(), n));
+                return operands.length == 0 ? null : mcb.and(operands);
```

Comment:
During rebase I've debugged the code and at this point operands contains two identical instances of the same MCB, I was wondering if it is expected. It seems like bug to me, but maybe there is something wrong in my end. It ends up calling ModelCriteriaBuilder.and method where there are two identical builders. I'll perform some additional tests to investigate it further. In the meantime I wanted to let you know about it. Thank you

Reply:
@vramik, is this happening also after the latest changes from @hmlnarik? I can imagine it was caused by not copying the tree in `and` and `or` methods, but it is not there anymore.

Reply:
@mhajas it is happening with JpaClientModelCriteriaBuilder from jpa store work and thanks to @hmlnarik it turned out it is caused by the fact I was returning `this` there. We've discussed it and now I'm going to rewrite my code. So this comment chould be resolved.